File: app/src/main/java/com/trindade/gamide/ui/fragments/EditorFragment.java
package com.trindade.gamide.ui.fragments;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.trindade.gamide.R;
import com.trindade.gamide.MainActivity;

import gslang.interpreter.BaseCompiller;

import com.google.android.material.transition.MaterialSharedAxis;

public class EditorFragment extends Fragment {
	
	// GamIDE Classes
	BaseCompiller baseCompiler;
	
	private Context mCtx;
	private EditText codeEditText;
	private Button runCode;
	
	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setEnterTransition(new MaterialSharedAxis(MaterialSharedAxis.X, true));
		setExitTransition(new MaterialSharedAxis(MaterialSharedAxis.X, false));
	}
	
	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.fragment_editor, container, false);
		mCtx = getContext(); // Inicialize o contexto aqui
		
		// GamIDE
		baseCompiler = new BaseCompiller(mCtx, MainActivity.class);
		
		codeEditText = view.findViewById(R.id.code);
		runCode = view.findViewById(R.id.runCode);
		
		try{
			runCode.setOnClickListener(v -> {
				baseCompiler.compile(codeEditText.getText().toString());
			});
		} catch (Exception e) {
			Toast.makeText(mCtx, e.toString(), 4000).show();
		}
		
		return view;
	}
	
	@Override
	public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
	}
}

File: app/src/main/java/dev/trindadeaquiles/robokide/ui/fragments/editor/EditorFragment.kt
package dev.trindadeaquiles.robokide.ui.fragments.editor

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout

import com.google.android.material.navigation.NavigationView
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.transition.MaterialSharedAxis
import com.google.android.material.tabs.TabLayout

import dev.trindadeaquiles.robokide.R
import dev.trindadeaquiles.robokide.databinding.FragmentEditorBinding
import dev.trindadeaquiles.robokide.ui.terminal.RobokTerminal
import dev.trindadeaquiles.robokide.ui.base.RobokFragment
import dev.trindadeaquiles.robokide.ui.components.progress.DotProgressBar
import dev.trindadeaquiles.robokide.ui.fragments.build.output.OutputFragment
import dev.trindadeaquiles.robokide.ui.fragments.editor.logs.LogsFragment
import dev.trindadeaquiles.robokide.ui.fragments.editor.diagnostic.DiagnosticFragment

import robok.dev.compiler.logic.LogicCompiler
import robok.dev.compiler.logic.LogicCompilerListener
import robok.dev.diagnostic.logic.DiagnosticListener

class EditorFragment(private val transitionAxis: Int = MaterialSharedAxis.X) : RobokFragment(transitionAxis) {

    private var _binding: FragmentEditorBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _binding = FragmentEditorBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        
        /*configureToolbar()*/
        
        val path = arguments?.getString(PROJECT_PATH) ?: "/sdcard/Robok/Projects/Default/"

        val terminal = RobokTerminal(requireContext())

        val compilerListener = object : LogicCompilerListener {
            override fun onCompiling(log: String) {
                terminal.addLog(log)
            }

            override fun onCompiled(output: String) {
                val outputFragment = OutputFragment(MaterialSharedAxis.X)
                outputFragment.addOutput(requireContext(), layoutInflater, view as ViewGroup, output)

                Snackbar.make(binding.root, R.string.message_compiled, Snackbar.LENGTH_LONG)
                    .setAction(R.string.go_to_outputs) {
                        openFragment(outputFragment)
                        terminal.dismiss()
                    }
                    .show()
            }
        }
        
        val diagnosticListener = object : DiagnosticListener {
            override fun onDiagnosticReceive (status: Boolean, diagnostic: String) {
                if (status) {
                   setDiagnosticOk()
                } else {
                   setDiagnosticError()
                }
            }
        }

        val compiler = LogicCompiler(requireContext(), compilerListener)

        binding.runButton.setOnClickListener {
            val code = binding.codeEditor.text.toString()
            terminal.show()
            compiler.compile(code)
        }

        binding.seeLogs.setOnClickListener {
            terminal.show()
        }

        tabLayoutConfig()
    }

    private fun tabLayoutConfig() {
        binding.tabLayout.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab?) {
                tab?.let {
                    when (it.text) {
                        getString(R.string.text_logs) -> {
                            openCustomFragment(R.id.drawer_editor_fragment_container, LogsFragment(MaterialSharedAxis.Y))
                        }

                        getString(R.string.text_diagnostic) -> {
                            openCustomFragment(R.id.drawer_editor_fragment_container, DiagnosticFragment(MaterialSharedAxis.Y))
                        }
                        else -> {
                            // Handle other tabs if needed
                        }
                    }
                }
            }

            override fun onTabReselected(tab: TabLayout.Tab?) { }
            override fun onTabUnselected(tab: TabLayout.Tab?) { }
        })
    }
    
    fun setDiagnosticError () {
    
    }
    
    fun setDiagnosticOk () {
    
    }
    
    fun configureToolbar() {
        binding.toolbar.setTitleCentered(false)
        val dotProgressBar = DotProgressBar.Builder()
              .setMargin(4)
              .setAnimationDuration(2000)
              .setMaxScale(1f)
              .setMinScale(0.3f) 
              .setNumberOfDots(3)
              .setDotRadius(8)
              .build(requireContext())
        binding.toolbar.addView(dotProgressBar)      
        dotProgressBar.startAnimation()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    companion object {
        private const val PROJECT_PATH = "arg_path"

        fun newInstance(path: String): EditorFragment {
            return EditorFragment(MaterialSharedAxis.X).apply {
                arguments = Bundle().apply {
                    putString(PROJECT_PATH, path)
                }
            }
        }
    }
}